### DISCUSSION TITLE
HTTP streaming response support for Azure functions [In Public Preview Now]

### DISCUSSION THREAD

**Discussion by nitinl** [Azure/azure-functions-python-worker#1349] [General]
Is there support for real-time streaming HttpResponse from Azure function?

This question has been asked [before](https://learn.microsoft.com/en-us/answers/questions/1334227/how-to-send-out-the-azureopen-ai-response-in-real), and the solution looks exactly like what is needed. However, func.HttpResponse object has no attribute 'write'.

**Comment by YunchuWang** [comment]
@nitinl thanks for following up! we are releasing a new worker version which supports http streaming req/response soon. stay tuned!

**Reply by jjczopek** [reply]
@YunchuWang any updates on that?

**Reply by samail** [reply]
+1 :)

**Reply by safaksaylam** [reply]
+1

**Reply by ghost** [reply]
+1

**Reply by bastienjalbert** [reply]
+10

**Reply by RomainFALLER** [reply]
+10. Already the 2nd project on which the question has arisen. That would be really nice. :/

**Reply by Anton-Buzik** [reply]
+1

**Reply by alitilitex** [reply]
+1

**Reply by wl02638153** [reply]
+1

**Reply by inrock1** [reply]
+1

**Reply by gautamsabba** [reply]
+1

**Reply by rgn15996** [reply]
Any updates on ETA? Everyone wants to return nice streamed responses from LLMs!

**Reply by zhoxing-ms** [reply]
+1

**Reply by RoyceAroc** [reply]
+1

**Reply by DJSaunders1997** [reply]
+1

**Reply by eavanvalkenburg** [reply]
+1

**Comment by bastienjalbert** [comment]
Any news about current implementation status ?

**Comment by tyler-suard-parker** [comment]
@YunchuWang Sorry to bother you about this again, but my company's project is looking to stream response from an Azure functions app.  Is that supported yet?

**Reply by fabge** [reply]
while you might have to wait a while until this is implemented, you could actually just use Azure Web PubSub as intermediary and stream responses this way.
e.g. a streaming response from the server could look like this:

1. user connects to websocket (Azure Web PubSub) using a /connect endpoint
2. user sends a message to /message endpoint
3. azure function receives the message and uses the azure pub sub sdk ([e.g. for python](https://learn.microsoft.com/en-us/python/api/overview/azure/messaging-webpubsubservice-readme?view=azure-python)) to stream the response via Azure Web PubSub to the user
4. user receives the message via the existing websocket connection

**Comment by waynebaby** [comment]
I am using .net, and i am confusing if i can use streaming output in .net version or it will ship with python features together?

**Reply by bastienjalbert** [reply]
It seems to be working with .net function already

**Reply by westleyMS** [reply]
Having issues with .net function streaming, seems I get response buffering causing issues. Can you point a link to a sample?

**Comment by RoyceAroc** [comment]
when is the expected release date?

**Reply by andredewes** [reply]
@YunchuWang Is there any rough estimate when this will be released?

**Comment by YunchuWang** [comment]
Hi everyone, thank you for your attention. We are actively working on supporting this feature for python function as one of top priorities.
The latest update on timeline is as below:

- Prototyping/Design completed internally
- PRs merged and ready for release.
- Public preview: [around Microsoft build event time (May 21-23)](https://azure.microsoft.com/en-us/updates/public-preview-azure-functions-support-for-http-streams-in-python/)

If there is any update on the timeline, i will share it here.

update1: feature is being released and will be in next azure function runtime release today
update2: we are preparing for public preview once released and will be out soon. will keep you updated.
update3: no need to sign up and instructions to enable http streaming will be given in public preview announcements along with samples
update4: it will be released into public preview around Microsoft build event time (May 21-23)
update5:
Hi everyone, thank you for the attention and interest. Great news, the http streaming feature is released in public preview today! Please feel free to check it out following the announcement https://azure.microsoft.com/en-us/updates/public-preview-azure-functions-support-for-http-streams-in-python/ and let us know your experience/feedback/issues :)!

**EDIT:** HTTP streaming is GA now! https://azure.microsoft.com/en-us/updates?id=492057

**Reply by dslam** [reply]
how to sign up for private review?

**Reply by DJSaunders1997** [reply]
Can I sign up for private preview please?

**Reply by safaksaylam** [reply]
How to sign up for private review?

**Reply by YunchuWang** [reply]
right now we are working on the details. Will release private preview signup instructions later when its available. thanks for your attention and support.

**Reply by sirishwizr** [reply]
Any update on this one. How to register for Private preview?

**Reply by juanpabloguerra16** [reply]
how long until this gets implemented? How can we sign up for preview?

**Reply by mgburnsy** [reply]
Any update on this?

**Reply by frederikruelens** [reply]
Hi Yunchu, Could you share any recent updates regarding the release of this feature? Thanks!

**Reply by alan-czysz** [reply]
Any updates for the timeline? Really looking forward to using streaming responses!

**Reply by habibregask86** [reply]
released preview/beta version

**Reply by mycaule** [reply]
I get this error while running the extension with Python 3.11 on a macOS Intel.
and latest versions of 

```
azure-functions = "^1.20.0"
azurefunctions-extensions-http-fastapi = "^1.0.0b1"
```

```txt
Exception: FunctionLoadError: cannot load the stream_count function: 
'req' binding type "httpTrigger" and dataType "None" in function.json 
do not match the corresponding function parameter's Python type annotation "Request"
```

What should I put in `function.json` ?

**Reply by jvdalen** [reply]
@mycaule I'm having the same issue. After a lot of testing it seems that once you import 
`from azurefunctions.extensions.http.fastapi import Request, StreamingResponse`

you MUST use req: Request and not req: Func.httprequest and replace all with the fastapi versions of the functions. 

I can't seem to find a way to mix and match the fastapi and azurefunction request types in one function app, and am not good enough of a coder to understanding how the bindings work and why that's an issue. Love for someone to clarify.

**Reply by mycaule** [reply]
Thanks, finally I didn't use `StreamingResponse` from FastAPI, because of my syntax problems, but also lack of support in the main Azure Functions package.

What I did is I implemented a WebSocket server somewhere else in my architecture, not running with App Services, and sent websockets using the [`python-socketio emit()`](https://python-socketio.readthedocs.io/en/latest/client.html#emitting-events) function. There are sample architectures in the documentation using _Azure Web PubSub_, but this is essentially the same as running a regular websocket server on App Service, at the difference you have to handle the hosting and scaling yourself.

I would recommend websockets for someone who needs a real time architecture.

**Reply by linde12** [reply]
One eternity later... The (still, as of Feb 2025) beta of `azurefunctions-extensions-http-fastapi` still requires you to replace all `func.HttpRequest` to `Request` and it is also using fixed versions of pydantic so if your project uses pydantic you will have to settle with an outdated variant because of this extension package.

How is this even an acceptable solution for such a basic function?

Adding `azurefunctions-extensions-http-fastapi` pulls in not only `fastapi` but also `anyio`, `pymongo`, `click`, `uvicorn` and a bunch of other dependencies. Why do i need all these dependencies for such a basic feature?

Why not expose a `write` method or even better, allow the `body` to be a generator function (similar to the Node SDK where body can be a `ReadStream`)? Why does this have to be an extension, especially one relying on fastapi?

Great, i can store files in Azure Blob Storage but i cannot send them to the client because i would have to load gigabytes of data into memory before i construct & return a `func.HttpResponse`, instead of streaming chunks... :+1:

**Reply by shahztechfarm** [reply]
I am facing the exact same issue. When I import 
`from azurefunctions.extensions.http.fastapi import Request, StreamingResponse
`
I get the error 
```
[2025-02-17T18:13:55.941Z] Indexed function app and found 4 functions
[2025-02-17T18:13:55.945Z] Worker failed to index functions
[2025-02-17T18:13:55.946Z] Result: Failure
Exception: FunctionLoadError: cannot load the Function01 function: 'req' binding type "httpTrigger" and dataType "None" in function.json do not match the corresponding function parameter's Python type annotation "HttpRequest"
```

**Reply by yashchks87** [reply]
Same issue as mentioned above. Can you give update on this?

**Reply by s174182** [reply]
@yashchks87 I got the same experience with the same error message, it works for me if i switch my app from FastApi to  eg:
`
import azure.functions as func
from FastApiApp import app as fastapi_app
app = func.AsgiFunctionApp(app=fastapi_app, http_auth_level=func.AuthLevel.ANONYMOUS)
´

To:
`
import azure.functions as func
from azurefunctions.extensions.http.fastapi import Request, StreamingResponse
import asyncio
import os
app = func.FunctionApp(http_auth_level=func.AuthLevel.ANONYMOUS)
´

**Reply by s174182** [reply]
Also, it seems that only "func.FunctionApp" supports actual streaming. func.AsgiFunctionApp still waits for the streaming to complete before returning a response.

Can someone elaborate whether above is correct and why such might be the case?

**Comment by vasudevan-n** [comment]
@YunchuWang  Can i know if its supported for FunctionApps written in PowerShell ?
